temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandlerBase.java

```diff
@@ -50,24 +51,12 @@ public static <T> T newProxy(Class<T> activityInterface, InvocationHandler invoc
             invocationHandler);
   }
 
-  protected void init(Class<?> activityInterface) {
-    POJOActivityInterfaceMetadata activityMetadata =
-        POJOActivityInterfaceMetadata.newInstance(activityInterface);
-    for (POJOActivityMethodMetadata methodMetadata : activityMetadata.getMethodsMetadata()) {
-      Method method = methodMetadata.getMethod();
-      String activityType = methodMetadata.getActivityTypeName();
-      MethodRetry methodRetry = method.getAnnotation(MethodRetry.class);
-      Function<Object[], Object> function = getActivityFunc(method, methodRetry, activityType);
-      methodFunctions.put(method, function);
-    }
-  }
-
   @Override
   public Object invoke(Object proxy, Method method, Object[] args) {
-    Function<Object[], Object> function = methodFunctions.get(method);
-    if (function == null) {
-      throw new IllegalArgumentException("Unexpected method: " + method);
-    }
+    POJOActivityMethodMetadata methodMetadata = activityMetadata.getMethodMetadata(method);
+    MethodRetry methodRetry = methodMetadata.getMethod().getAnnotation(MethodRetry.class);
+    String activityType = methodMetadata.getActivityTypeName();
+    Function<Object[], Object> function = getActivityFunc(method, methodRetry, activityType);
```

Comment:
Could memoize this is if it's frequent/expensive enough

Reply:
That will be a separate PR. The main thing that we should memoize/cache is all this crazy reflection we are doing inside POJO*Metadata classes again and again for the same classes and interfaces.